integration/container/mounts_linux_test.go

```diff
@@ -265,3 +267,36 @@ func TestContainerBindMountNonRecursive(t *testing.T) {
 		poll.WaitOn(t, container.IsSuccessful(ctx, client, c), poll.WithDelay(100*time.Millisecond))
 	}
 }
+
+func TestMountSubPath(t *testing.T) {
+	skip.If(t, testEnv.IsRemoteDaemon)
+
+	defer setupTest(t)()
+
+	ctx := context.Background()
+	d := daemon.New(t)
+	client := d.NewClientT(t)
+
+	d.StartWithBusybox(t)
+	defer d.Stop(t)
+
+	volumeName := t.Name() + "_vol_foo"
+	_, err := client.VolumeCreate(context.Background(), volumetypes.VolumeCreateBody{Name: volumeName})
+	assert.NilError(t, err)
+
+	mnt := mounttypes.Mount{
```

Comment:
We might need some more test-cases (probably good to combine them in a single test, using subtests). Thinking of;


- A container using the same volume _twice_ but with different sub-paths (also to make sure reference-counting works correctly (although I'm not sure if that makes a difference - just a quick thought)
- A test where a non-existing sub-path is used (i.e., make sure if produces an error)
- Multiple containers using the same volume + sub-paths? (not sure if needed; just thinking out loud; is reference-counting the same?)
- What happens if the volume is used twice (either the same container or different containers), and the chosen sub-paths lead to "nested mounts"? (i.e. mount sub-path `/foo/bar/baz` and `/foo/bar` ?)
  - does this lead to problems, such as "unable to unmount" if one container is stopped?
  - what happens with `NoCopy=false` if the container that uses `/foo/bar` happens to have a subdirectory (or file) named `/foo/bar/baz` ?


Perhaps there's more/other (corner) cases that we should check for.  


edit: I just noticed you have a test for the non-existent path below; still might be worth looking for some additional cases 😅

Reply:
Thanks, I will look into adding more test cases (looks like most of these cases need integration tests), while waiting for more review comments.

Reply:
- I have added a test for one container mounting the same volume twice, with different sub-paths. I think a single container will count as 1 reference to the volume, regardless of how many times it mounts the volume using different sub-paths, since the reference counting for each volume is based on container ID.

- Non-existent sub-path test, as you mentioned in the comment, is covered by unit tests (I added an integration test too, but perhaps it can be removed once it has served its purpose of verifying the propagation of non-existent sub-path error)

- With the current implementation proposed by this PR, the sub-path logic is in the `MountPoint` layer, and thus will not affect the reference counting (volume name + container id combination), i.e. mounting the same volume in different containers will have the same effect on reference counts with or without sub-path option (please let me know if this requires an explicit integration test).

Regarding overlap of sub-paths:
- With one or multiple containers mounting a single volume, I could not find a scenario where overlap of sub-paths would affect ability to un-mount the volume, as the the reference count remains the same. (unless there is another mechanism/aspect that I have missed). The PR now includes an integration test for verifying.

- Copying contents from the FS of a container’s image into a volume currently requires the to-be-populated volume to be empty, i.e. there won’t be an overlap between the image’s FS and the specified sub-dir. The checks performed on mounts, prior to any copying to volumes, includes verifying the existence of the specified sub-path. (I have added an integration test to warn against possible changes to this behaviour).